src/common/src/array/arrow/arrow_impl.rs

```diff
@@ -854,21 +956,35 @@ converts!(Utf8Array, arrow_array::StringArray);
 converts!(Utf8Array, arrow_array::LargeStringArray);
 converts!(DateArray, arrow_array::Date32Array, @map);
 converts!(TimeArray, arrow_array::Time64MicrosecondArray, @map);
-converts!(TimestampArray, arrow_array::TimestampMicrosecondArray, @map);
-converts!(TimestamptzArray, arrow_array::TimestampMicrosecondArray, @map);
+
+converts!(TimestampArray, arrow_array::TimestampSecondArray, TimeUnit::Second, @map);
+converts!(TimestampArray, arrow_array::TimestampMillisecondArray, TimeUnit::Millisecond, @map);
+converts!(TimestampArray, arrow_array::TimestampMicrosecondArray, TimeUnit::Microsecond, @map);
+converts!(TimestampArray, arrow_array::TimestampNanosecondArray, TimeUnit::Nanosecond, @map);
+
+converts!(TimestamptzArray, arrow_array::TimestampSecondArray, TimeUnit::Second, @map);
+converts!(TimestamptzArray, arrow_array::TimestampMillisecondArray,TimeUnit::Millisecond, @map);
+converts!(TimestamptzArray, arrow_array::TimestampMicrosecondArray, TimeUnit::Microsecond, @map);
+converts!(TimestamptzArray, arrow_array::TimestampNanosecondArray, TimeUnit::Nanosecond, @map);
+
 converts!(IntervalArray, arrow_array::IntervalMonthDayNanoArray, @map);
 converts!(SerialArray, arrow_array::Int64Array, @map);
 
 /// Converts RisingWave value from and into Arrow value.
 trait FromIntoArrow {
     /// The corresponding element type in the Arrow array.
     type ArrowType;
+    /// The timestamp type used to distinguish different time units, only utilized when the Arrow type is a timestamp.
+    type TimestampType;
     fn from_arrow(value: Self::ArrowType) -> Self;
     fn into_arrow(self) -> Self::ArrowType;
+    /// Used for converting timestamp types and will not be used in conversions of other types.
+    fn from_arrow_with_unit(value: Self::ArrowType, time_unit: Self::TimestampType) -> Self;
 }
```

Comment:
Generally speaking, for type conversion, schema (i.e., type information) should be provided, since data itself may not be self-contained.

```rust
fn convert(v_from, t_from, t_to) -> v_to 
```
`t_from` & `t_to` are not always needed though

e.g., our Avro conversion

https://github.com/risingwavelabs/risingwave/blob/9a32e75f227f837a6f16fb3afe159c1f8af2ad5b/src/connector/codec/src/decoder/avro/mod.rs#L89-L93

Reply:
Therefore, `_with_unit` looks a little strange to me. Haven't checked the details of Arrow yet.

Reply:
cc @xiangjinwu @BugenZhao might also be able to comment

Reply:
https://docs.rs/arrow-schema/53.2.0/arrow_schema/enum.DataType.html
If we need `_with_unit` to pass extra info from `DataType`, how about the following?
* `IntervalUnit`
* `i32` length for `FixedSizeBinary`
* precision and scale for `Decimal{128,256}`

To repeat, `DataType` should be provided as data itself may not be self-contained.